Improve the caching strategy employed in `utils.py`

AlexWaygood opened the pull request:
Various hot functions in `utils.py` are cached, since they are called with the same arguments from multiple hooks in `checkers.py`. However, the current cache size is far lower than it needs to be, at 128; many `.rst` files in CPython have more than 128 paragraphs in them. It's also inefficient, since the cache entries from previous files are retained when checking a new file, but it's very rare that two different docs files have identical paragraphs between them.

We can improve the caching strategy by changing the caches so that they are per-file caches: the cache is allowed to grow without limit while checking any one file, but is cleared after the file has been checked.

This PR cuts around 50% off the time it takes `sphinx-lint` to check cpython's `Doc/` directory on my machine.

Part of #76

AlexWaygood commented:
Many thanks to @ezio-melotti for pointing out that the current caching strategy was probably woefully inefficient, and to @hugovk for coming up with the idea of clearing the caches after checking every file!

hugovk commented:
Wow! For me, 1.681s -> 1.004s on CPython docs = 40% faster! 🚀🚀

hugovk commented on sphinxlint/utils.py:
```diff
@@ -7,6 +7,15 @@
+PER_FILE_CACHES = []
+
+
+def _per_file_cache(func):
```
We have no API, remove the underscore?
```suggestion
def per_file_cache(func):
```

AlexWaygood replied:
Sure!

AlexWaygood commented:
In https://github.com/sphinx-contrib/sphinx-lint/pull/84/commits/715ffb232ad4eba0e9f491e8191cf9c9760cf1c2, I removed a more complex cache that I added in https://github.com/AlexWaygood/sphinx-lint/commit/7d63ef34ecc9d594692f6b39974a43e9b565414b. I think any speedup I measured from adding that cache must have been a total fluke. Local experiments indicated that the cache never had any successful hits, and since it was an unbounded cache, it was pretty problematic in terms of memory consumption. I can't measure any slowdown from removing it now.

hugovk reviewed (approved):
Similar timings, 1.606->0.998s 👍

AlexWaygood commented:
I just checked all the remaining caches. They all look useful to me: they all either have a huge amount of hits after running on a few `.rst` files in cpython, or they have a very good hit/miss ratio.
